README.md
# NAME

Validator::Procedural - Procedural validator

# SYNOPSIS

    use Validator::Procedural;

    # create validator prototype with given filters / checkers.
    my $prot = Validator::Procedural::Prototype->new(
        filters => {
            UCFIRST => sub { ucfirst },
        },
        checkers => {
            NUMERIC => sub { /^\d+$/ || 'INVALID' },
        },
    );

    # filter plugins can be applied to validator (prototypes).
    Validator::Procedural::Filter::Common->register_to($prot, 'TRIM', 'LTRIM');
    # also checker plugins can be
    Validator::Procedural::Checker::Common->register_to($prot);

    # filter class registration can be called from validator (prototypes).
    # package begins with '::' is recognized under 'Validator::Procedural::Filter::' namespace.
    $prot->register_filter_class('::Japanese', 'HAN2ZEN');
    $prot->register_filter_class('MY::Own::Filter');
    # also for checker class registration
    # package begins with '::' is recognized under 'Validator::Procedural::Checker::' namespace.
    $prot->register_checker_class('::Date');
    $prot->register_checker_class('MY::Own::Checker', 'MYCHECKER');

    # you can register filters (and checkers) after instantiation of prototype
    $prot->register_filter(
        TRIM => sub { s{ (?: \A \s+ | \s+ \z ) }{}gxmso; $_ },
    );

    $prot->register_checker(
        EMAIL => sub {
            # Of course this is not precise for email address, but just example.
            unless (m{\A \w+ @ \w+ (?: \. \w+ )+ \z}xmso) {
                return 'INVALID';   # error code for errors
            }

            return;                 # (undef) for OK
        },
    );

    # can register common filtering and checking procedure (not required)
    $prot->register_procedure('DATETIME', sub {
        my ($field) = @_;

        # apply filters
        $field->apply_filter('TRIM');

        # apply checkers
        return unless $field->check('EXISTS');
    });

    # register error message formatter (default is Validator::Procedural::Formatter::Minimal)
    $prot->register_formatter(Validator::Procedural::Formatter::Minimal->new());

    # now create validator (with state) from prototype
    my $validator = $prot->create_validator();

    # process for field 'bar' by custom procedure
    $validator->process('bar', sub {
        my ($field) = @_;

        # set value
        $field->value($req->param('bar'));

        # apply filters
        $field->apply_filter('TRIM');

        # filter value manually
        my $val = $field->value();
        $field->value($val . ' +0000');

        # apply checkers
        return unless $field->check('EXISTS');

        # check value manually
        eval {
            require Time::Piece;
            Time::Piece->strptime($field->value, '%Y-%m-%d %z');
        };
        if ($@) {
            $field->add_error('INVALID_DATE');
            return;
        }
    });

    # can apply registered procedure with given value
    $validator->process('foo', 'DATETIME', $req->param('foo'));



    # can retrieve validation result anytime

    $validator->success();      # => TRUE or FALSE
    $validator->has_error();    # => ! success()

    # retrieve fields and errors mapping
    $validator->errors();       # return errors errors in Array or Hash-ref (for scalar context)
    # => (
    #     foo => [ 'MISSING', 'INVALID_DATE' ],
    # )

    $validator->invalid_fields();
    # => ( 'foo', 'bar' )

    # can filter fields that has given error code
    $validator->invalid_fields('MISSING');
    # => ( 'foo', 'bar' )

    # error code filtering rule can be supplied with subroutine
    $validator->invalid_fields(sub { grep { $_ eq 'MISSING' } @_ });
    # => ( 'foo', 'bar' )

    $validator->valid('foo');       # => TRUE of FALSE
    $validator->invalid('foo');     # => ! valid()

    # retrieve error codes (or empty for valid field)
    $validator->error('foo');       # return errors for specified field in Array
    # => ( 'MISSING', 'INVALID_DATE' )

    # clear all errors
    $validator->clear_errors();
    # clear errors for specified fields
    $validator->clear_errors('foo');

    # append error (manually)
    $validator->add_error('foo', 'MISSING');

    # retrieve filtered value for specified field
    $validator->value('foo');
    # retrieve all values filtered
    $validator->values();   # return values in Array or Hash::MultiValue (for scalar context)
    # => (
    #     foo => [ 'val1', 'val2' ],
    #     var => [ 'val1' ],            # always in Array-ref for single value
    # )

    # retrieve error messages for all fields
    $validator->error_messages();
    # retrieve error message(s) for given field
    $validator->error_message('foo');

# DESCRIPTION

Validator::Procedural is yet another validation module.

THIS MODULE IS CURRENTLY ON WORKING DRAFT PHASE.  API MAY CHANGE.

# MOTIVATION FOR YET ANOTHER VALIDATION MODULE

There are so many validation modules on CPAN.  Why yet another one?

Some of modules provide good-looking features with simple configuration, but when I used those modules for compositing several fields and filtering fields (and for condition of some fields depending on other field), some were not able to handle such situation, some required custom handler.

So I focused on following points for design this module.

- To provide compact but sufficient container for validation result
- To provide filtering mechanism and functionality to retrieve filtered parameters
- To depend on other modules as least as possible (complex validators and filters depending on other modules heavyly will be supplied as dependent plugin distributions)

This module is NOT all-in-one validation product.  This module DOES NOT provide easy configuration.  But you have to implement validation procedure with Perl code, so on such a complex condition described above, you can make codes straightforwardly, easy to understand.

# LICENSE

Copyright (C) ITO Nobuaki.

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

# AUTHOR

ITO Nobuaki <dayflower@cpan.org>


Changes
Revision history for Perl extension Validator-Procedural

{{$NEXT}}

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Validator-Procedural',
    module_name     => 'Validator::Procedural',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Validator/Procedural/Field.pm
package Validator::Procedural::Field;
use 5.008005;
use strict;
use warnings;

use Validator::Procedural;

1;
__END__

=encoding utf-8

=head1 NAME

Validator::Procedural::Field - Field for Validator::Procedural

=head1 SYNOPSIS

    $validator->process('bar', sub {
        my ($field) = @_;

        # set value
        $field->value($req->param('bar'));

        # apply filters
        $field->apply_filter('TRIM');

        # filter value manually
        my $val = $field->value();
        $field->value($val . ' +0000');

        # apply checkers
        return unless $field->check('EXISTS');

        # check value manually
        eval {
            require Time::Piece;
            Time::Piece->strptime($field->value, '%Y-%m-%d %z');
        };
        if ($@) {
            $field->add_error('INVALID_DATE');
            return;
        }
    });

=head1 DESCRIPTION

This is the field class of L<Validator::Procedural>.

You will use instances of this class in validation "procedure"s, so you do not have to C<use> L<Validator::Procedural::Field> directly.

=head1 METHODS

=over 4

=item label

Returns field label.

=item value

    my $val = $field->value();

    $field->value(undef);           # unset value for field
    $field->value('foo');
    $field->value('foo', 'bar');    # can set multiple values for field

Gets and sets field value.

=item apply_filter

    $field->apply_filter('TRIM');           # apply registered filter
    $field->apply_filter(sub { ... });
    $field->apply_filter('X', times => 3);  # can supply options for filter

Applies filter to field value.

=item check

    $field->check('TRIM');              # call registered checker
    $field->check(sub { ... });
    $field->check('GT', than => 3);     # can supply options for checker

Checks field value.

Returns C<1> (true) for success, C<0> (false) for failure.
So you can stop further validations on failure by checking result.
Also you can continue validations.

=item error

Returns current error codes for the field.

=item add_error

    $field->add_error('INVALID');
    $field->add_error('UGLY', 'BAD');   # can supply multiple codes

Appends error code(s) for the field.

=item clear_errors

Clears error codes for the field.

=item set_errors

    $field->set_error('INVALID');
    $field->set_error('UGLY', 'BAD');   # can supply multiple codes

Sets error code(s) for the field.

And clears current error codes, so this is discouraged to use for normal conditions.

=back

=head1 LICENSE

Copyright (C) ITO Nobuaki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

ITO Nobuaki E<lt>dayflower@cpan.orgE<gt>

=cut



lib/Validator/Procedural/Prototype.pm
package Validator::Procedural::Prototype;
use 5.008005;
use strict;
use warnings;

use Validator::Procedural;

1;
__END__

=encoding utf-8

=head1 NAME

Validator::Procedural::Prototype - Prototype for Validator::Procedural

=head1 SYNOPSIS

    use Validator::Procedural;

    my $prot = Validator::Procedural::Prototype->new(
        filters => {
            UCFIRST => sub { ucfirst },
        },
        checkers => {
            NUMERIC => sub { /^\d+$/ || 'INVALID' },
        },
    );

    $prot->register_filter('FOO', sub { ... });

    $prot->register_checker_class('::Common');

    my $validator = $prot->create_validator();

=head1 DESCRIPTION

This is a prototype class for validator instances.

In web applications for example, it is recommended to create validator prototype once for the application startup, and to generate validator instances from the prototype for each request.

=head1 METHODS

=over 4

=item new

    my $prot = Validator::Procedural::Prototype->new(
        filters => {
            FOO => sub { ... },
        },
        checkers => {
            BAR => sub { ... },
        },
        procedures => {
            baz => sub { ... },
        },
    );

Creates prototype.

Optionally filters / checkers / procedures can be specified.

=item create_validator

Generates L<Validator::Procedural> instance from the prototype.

Registerd filter / checker / procedure methods in the prototype will be inherited by generated instances.

Doesn't consume any arguments.  (in current API)

=item register_filter

    $prot->register_filter( FOO => sub { ... } );
    $prot->register_filter(
        FOO => sub { ... },
        BAR => sub { ... },
        ...
    );

Registers filter methods.

Requisites for filter methods are described in L<Validator::Procedural::Filter>.

=item register_checker

    $prot->register_checker( FOO => sub { ... } );
    $prot->register_checker(
        FOO => sub { ... },
        BAR => sub { ... },
        ...
    );

Registers checker methods.

Requisites for checker methods are described in L<Validator::Procedural::Checker>.

=item register_procedure

    $prot->register_procedure( FOO => sub { ... } );
    $prot->register_procedure(
        FOO => sub { ... },
        BAR => sub { ... },
        ...
    );

Registers procedure methods.

Requisites for procedure methods are described in L<Validator::Procedural::Procedure>.

=item register_filter_class

    # register filter methods of Validator::Procedural::Filter::Common
    $prot->register_filter_class('::Common');

    $prot->register_filter_class('MY::Own::Filter::Class');

    # restrict registering methods (like Perl's importer)
    $prot->register_filter_class('::Text', 'TRIM', 'LTRIM');

Register filter methods from specified module.
(Modules will be loaded automatically.)

=item register_checker_class

    # register checker methods of Validator::Procedural::Checker::Common
    $prot->register_checker_class('::Common');

    $prot->register_checker_class('MY::Own::Checker::Class');

    # restrict registering methods
    $prot->register_checker_class('::Number', 'BIG', 'SMALL');

Register checker methods from specified module.
(Modules will be loaded automatically.)

=item register_procedure_class

    # register procedure methods of Validator::Procedural::Procedure::Common
    $prot->register_procedure_class('::Common');

    $prot->register_procedure_class('MY::Own::Procedure::Class');

    # restrict registering methods
    $prot->register_procedure_class('::Text', 'address', 'telephone');

Register procedure methods from specified module.
(Modules will be loaded automatically.)

=back

=head1 LICENSE

Copyright (C) ITO Nobuaki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

ITO Nobuaki E<lt>dayflower@cpan.orgE<gt>

=cut



lib/Validator/Procedural/Formatter/Minimal.pm
package Validator::Procedural::Formatter::Minimal;
use 5.008005;
use strict;
use warnings;

sub new {
    my ($class, %args) = @_;
    my $fields = delete $args{fields} || {};
    my $errors = delete $args{errors} || {};

    my $self = bless {
        fields => +{
            %$fields,
        },
        errors => +{
            MISSING => '[_1] is missing',
            INVALID => '[_1] is invalid',

            %$errors,
        },
    }, $class;

    return $self;
}

sub format {
    my ($self, $field, @error_codes) = @_;
    return unless @error_codes;

    my $field_name = $self->field_name($field);

    return map { s/\Q[_1]\E/$field_name/g; $_ }
           map { $self->error_template($_) }
               @error_codes;
}

sub error_template {
    my ($self, $error_code) = @_;

    unless (defined $self->{errors}->{$error_code}) {
        $self->{errors}->{$error_code} = $self->error_template_fallback($error_code);
    }

    return $self->{errors}->{$error_code};
}

sub error_template_fallback {
    my ($self, $error_code) = @_;
    return $self->{errors}->{INVALID};
}

sub field_name {
    my ($self, $field) = @_;

    unless (defined $self->{fields}->{$field}) {
        $self->{fields}->{$field} = $self->field_name_fallback($field);
    }

    return $self->{fields}->{$field};
}

sub field_name_fallback {
    my ($self, $field) = @_;

    return ucfirst lc $field;
}

1;
__END__

=encoding utf-8

=head1 NAME

Validator::Procedural::Formatter::Minimal - ...

=head1 SYNOPSIS

    use Validator::Procedural;
    use Validator::Procedural::Formatter::Minimal;

    my $formatter = Validator::Procedural::Formatter::Minimal->new(
        fields => {
            foo => 'Foo',
        },
        errors => {
            MISSING => '[_1] is missing',
            INVALID => '[_1] is invalid',
        },
    );

    my $prot = Validator::Procedural::Prototype->new(
        formatter => $formatter,
    );

    my $validator = $prot->create_validator();

=head1 DESCRIPTION

Validator::Procedural::Formatter::Minimal is ...

=head1 LICENSE

Copyright (C) ITO Nobuaki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

ITO Nobuaki E<lt>dayflower@cpan.orgE<gt>

=cut



lib/Validator/Procedural.pm
package Validator::Procedural;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.01";

package Validator::Procedural::_RegistryMixin;

use Carp;
use Module::Load ();

sub register_filter {
    my ($self, @args) = @_;
    while (my ($name, $filter) = splice @args, 0, 2) {
        $self->{filters}->{$name} = $filter;
    }
    return $self;
}

sub register_checker {
    my ($self, @args) = @_;
    while (my ($name, $checker) = splice @args, 0, 2) {
        $self->{checker}->{$name} = $checker;
    }
    return $self;
}

sub register_procedure {
    my ($self, @args) = @_;
    while (my ($name, $procedure) = splice @args, 0, 2) {
        $self->{procedure}->{$name} = $procedure;
    }
    return $self;
}

foreach my $prop (qw( filter checker procedure )) {
    my $prefix = ucfirst $prop;
    my $sub = sub {
        my ($self, $package) = splice @_, 0, 2;

        if ($package =~ s/^\:://) {
            # plugin namespace
            $package = "Validator::Procedural::$prefix::$package";
        }
        else {
            # full spec package name
        }

        Module::Load::load($package);

        $package->register_to($self, @_);

        return $self;
    };

    my $func = sprintf '%s::register_%s_class', __PACKAGE__, $prop;

    no strict 'refs';
    *{$func} = $sub;
}

package Validator::Procedural::Prototype;

our $VERSION = $Validator::Procedural::VERSION;

our @ISA = qw( Validator::Procedural::_RegistryMixin );

sub new {
    my ($class, %args) = @_;
    my $self = bless {
        filters    => {},
        checkers   => {},
        procedures => {},

        %args,
    }, $class;
    return $self;
}

sub register_formatter {
    my ($self, $formatter) = @_;
    $self->{formatter} = $formatter;
    return $self;
}

sub create_validator {
    my $self = shift;

    return Validator::Procedural->new(
        filters    => { %{$self->{filters}}    },
        checkers   => { %{$self->{checkers}}   },
        procedures => { %{$self->{procedures}} },
        formatter  => $self->{formatter},
    );
}

package Validator::Procedural;

use Carp;
use List::Util qw( first );
use Hash::MultiValue;

our @ISA = qw( Validator::Procedural::_RegistryMixin );

sub new {
    my ($class, %args) = @_;
    my $self = bless {
        filters    => {},
        checkers   => {},
        procedures => {},

        value        => {},
        value_fields => [],
        error        => {},
        error_fields => [],

        %args,
    }, $class;
    return $self;
}

sub formatter {
    my $self = shift;
    if (@_) {
        $self->{formatter} = shift;
    }

    unless ($self->{formatter}) {
        require Validator::Procedural::Formatter::Minimal;
        $self->{formatter} = Validator::Procedural::Formatter::Minimal->new();
    }

    return $self->{formatter};
}

sub process {
    my ($self, $field, $procedure, @vals) = @_;

    if (@vals) {
        $self->value($field, @vals);
    }

    if (! ref $procedure) {
        my $meth = $self->{procedures}->{$procedure};
        unless ($meth) {
            croak "Undefined procedure: '$procedure'";
        }
        $procedure = $meth;
    }

    &$procedure($self->field($field));

    return $self;
}

sub field {
    my ($self, $field_name) = @_;
    return Validator::Procedural::Field->new($self, $field_name);
}

sub value {
    my ($self, $field) = splice @_, 0, 2;

    if (@_) {
        if (@_ == 1 && ! defined $_[0]) {
            delete $self->{value}->{$field};
            @{$self->{value_fields}} = grep { $_ ne $field } @{$self->{value_fields}};
        }
        else {
            if (! exists $self->{value}->{$field}) {
                push @{$self->{value_fields}}, $field;
            }

            $self->{value}->{$field} = [ @_ ];
        }
    }

    return unless exists $self->{value}->{$field};
    my $vals = $self->{value}->{$field};
    return unless defined $vals;

    return wantarray ? @$vals : $vals->[0];
}

sub values {
    my ($self) = @_;

    my @values = map { $_ => $self->{value}->{$_} } @{$self->{value_fields}};
    return wantarray ? @values : Hash::MultiValue->from_mixed(@values);
}

sub apply_filter {
    my ($self, $field, $filter, %options) = @_;

    my @vals = $self->value($field);

    my $meth = $filter;
    if (! ref $meth) {
        $meth = $self->{filters}->{$filter};
        unless ($meth) {
            croak "Undefined filter: '$filter'";
        }
    }

    @vals = map { &$meth($_, \%options) } @vals;

    $self->value($field, @vals);

    return @vals;
}

sub check {
    my ($self, $field, $checker, %options) = @_;

    my @vals = $self->value($field);

    my $meth = $checker;
    if (! ref $meth) {
        $meth = $self->{checkers}->{$checker};
        unless ($meth) {
            croak "Undefined checker '$checker'";
        }
    }

    my @error_codes = do {
        local $_ = $vals[0];
        grep { defined $_ && $_ ne "" } &$meth(@vals, \%options);
    };
    if (@error_codes) {
        $self->add_error($field, @error_codes);
    }

    return @error_codes == 0;
}

sub success {
    my ($self) = @_;
    return ! $self->has_error();
}

sub has_error {
    my ($self) = @_;
    return @{$self->{error_fields}} > 0;
}

sub valid_fields {
    my ($self) = @_;

    my @fields = grep { ! exists $self->{error}->{$_} }
                      @{$self->{value_fields}};
    return @fields;
}

sub invalid_fields {
    my $self = shift;

    my @fields = @{$self->{error_fields}};

    my @crits = map { ref $_ ? $_ : _gen_invalid_matcher($_) } @_;
    if (@crits) {
        my $e = $self->{error};
        @fields = grep { my $f = $_; first { &$_(@{$e->{$f}}) } @crits }
                       @fields;
    }

    return @fields;
}

sub _gen_invalid_matcher {
    my ($error_code) = @_;
    return sub { first { $_ eq $error_code } @_ };
}

sub errors {
    my ($self) = @_;

    my @errors = map { $_ => $self->{error}->{$_} } @{$self->{error_fields}};
    return wantarray ? @errors : +{ @errors };
}

sub error {
    my ($self, $field) = splice @_, 0, 2;

    return unless exists $self->{error}->{$field};

    my @errors = @{$self->{error}->{$field}};

    return @errors;
}

sub valid {
    my ($self, $field) = @_;

    return ! $self->invalid($field);
}

sub invalid {
    my ($self, $field) = @_;

    return exists $self->{error}->{$field};
}

sub clear_errors {
    my ($self, $field) = @_;

    if (defined $field) {
        delete $self->{error}->{$field};
        @{$self->{error_fields}} = grep { $_ ne $field }
                                        @{$self->{error_fields}};
    }
    else {
        $self->{error} = {};
        $self->{error_fields} = [];
    }

    return $self;
}

sub set_errors {
    my ($self, $field) = splice @_, 0, 2;

    if (@_) {
        if (! exists $self->{error}->{$field}) {
            push @{$self->{error_fields}}, $field;
        }

        $self->{error}->{$field} = [ @_ ];
    }
    else {
        $self->clear_errors($field);
    }

    return $self;
}

sub add_error {
    my ($self, $field) = splice @_, 0, 2;

    if (! exists $self->{error}->{$field}) {
        push @{$self->{error_fields}}, $field;

        $self->{error}->{$field} = [];
    }

    push @{$self->{error}->{$field}}, @_;

    return $self;
}

sub error_messages {
    my ($self) = @_;

    my @messages = map { $self->error_message($_) } @{$self->{error_fields}};
    return @messages;
}

sub error_message {
    my ($self, $field) = @_;

    return $self->formatter->format($field, $self->error($field));
}

package Validator::Procedural::Field;

sub new {
    my ($class, $validator, $label) = @_;
    my $self = bless {
        validator => $validator,
        label     => $label,
    }, $class;
    return $self;
}

sub label { $_[0]->{label} }

foreach my $method (qw( value apply_filter check
                     error clear_errors set_errors add_error )) {
    my $sub = sub {
        my $self = shift;
        return $self->{validator}->$method($self->{label}, @_);
    };

    my $func = sprintf '%s::%s', __PACKAGE__, $method;

    no strict 'refs';
    *{$func} = $sub;
}

1;
__END__

=encoding utf-8

=head1 NAME

Validator::Procedural - Procedural validator

=head1 SYNOPSIS

    use Validator::Procedural;

    # create validator prototype with given filters / checkers.
    my $prot = Validator::Procedural::Prototype->new(
        filters => {
            UCFIRST => sub { ucfirst },
        },
        checkers => {
            NUMERIC => sub { /^\d+$/ || 'INVALID' },
        },
    );

    # filter plugins can be applied to validator (prototypes).
    Validator::Procedural::Filter::Common->register_to($prot, 'TRIM', 'LTRIM');
    # also checker plugins can be
    Validator::Procedural::Checker::Common->register_to($prot);

    # filter class registration can be called from validator (prototypes).
    # package begins with '::' is recognized under 'Validator::Procedural::Filter::' namespace.
    $prot->register_filter_class('::Japanese', 'HAN2ZEN');
    $prot->register_filter_class('MY::Own::Filter');
    # also for checker class registration
    # package begins with '::' is recognized under 'Validator::Procedural::Checker::' namespace.
    $prot->register_checker_class('::Date');
    $prot->register_checker_class('MY::Own::Checker', 'MYCHECKER');

    # you can register filters (and checkers) after instantiation of prototype
    $prot->register_filter(
        TRIM => sub { s{ (?: \A \s+ | \s+ \z ) }{}gxmso; $_ },
    );

    $prot->register_checker(
        EMAIL => sub {
            # Of course this is not precise for email address, but just example.
            unless (m{\A \w+ @ \w+ (?: \. \w+ )+ \z}xmso) {
                return 'INVALID';   # error code for errors
            }

            return;                 # (undef) for OK
        },
    );

    # can register common filtering and checking procedure (not required)
    $prot->register_procedure('DATETIME', sub {
        my ($field) = @_;

        # apply filters
        $field->apply_filter('TRIM');

        # apply checkers
        return unless $field->check('EXISTS');
    });

    # register error message formatter (default is Validator::Procedural::Formatter::Minimal)
    $prot->register_formatter(Validator::Procedural::Formatter::Minimal->new());

    # now create validator (with state) from prototype
    my $validator = $prot->create_validator();

    # process for field 'bar' by custom procedure
    $validator->process('bar', sub {
        my ($field) = @_;

        # set value
        $field->value($req->param('bar'));

        # apply filters
        $field->apply_filter('TRIM');

        # filter value manually
        my $val = $field->value();
        $field->value($val . ' +0000');

        # apply checkers
        return unless $field->check('EXISTS');

        # check value manually
        eval {
            require Time::Piece;
            Time::Piece->strptime($field->value, '%Y-%m-%d %z');
        };
        if ($@) {
            $field->add_error('INVALID_DATE');
            return;
        }
    });

    # can apply registered procedure with given value
    $validator->process('foo', 'DATETIME', $req->param('foo'));


    # can retrieve validation result anytime

    $validator->success();      # => TRUE or FALSE
    $validator->has_error();    # => ! success()

    # retrieve fields and errors mapping
    $validator->errors();       # return errors errors in Array or Hash-ref (for scalar context)
    # => (
    #     foo => [ 'MISSING', 'INVALID_DATE' ],
    # )

    $validator->invalid_fields();
    # => ( 'foo', 'bar' )

    # can filter fields that has given error code
    $validator->invalid_fields('MISSING');
    # => ( 'foo', 'bar' )

    # error code filtering rule can be supplied with subroutine
    $validator->invalid_fields(sub { grep { $_ eq 'MISSING' } @_ });
    # => ( 'foo', 'bar' )

    $validator->valid('foo');       # => TRUE of FALSE
    $validator->invalid('foo');     # => ! valid()

    # retrieve error codes (or empty for valid field)
    $validator->error('foo');       # return errors for specified field in Array
    # => ( 'MISSING', 'INVALID_DATE' )

    # clear all errors
    $validator->clear_errors();
    # clear errors for specified fields
    $validator->clear_errors('foo');

    # append error (manually)
    $validator->add_error('foo', 'MISSING');

    # retrieve filtered value for specified field
    $validator->value('foo');
    # retrieve all values filtered
    $validator->values();   # return values in Array or Hash::MultiValue (for scalar context)
    # => (
    #     foo => [ 'val1', 'val2' ],
    #     var => [ 'val1' ],            # always in Array-ref for single value
    # )

    # retrieve error messages for all fields
    $validator->error_messages();
    # retrieve error message(s) for given field
    $validator->error_message('foo');

=head1 DESCRIPTION

Validator::Procedural is yet another validation module.

THIS MODULE IS CURRENTLY ON WORKING DRAFT PHASE.  API MAY CHANGE.

=head1 MOTIVATION FOR YET ANOTHER VALIDATION MODULE

There are so many validation modules on CPAN.  Why yet another one?

Some of such modules provide good-looking features with simple configuration. But when I used those modules for compositing several fields and filtering fields (and for condition of some fields depending on other field), some were not able to handle such situation, some required custom handler.

So I focused on following points for design this module.

=over 4

=item To provide compact but sufficient container for validation result

=item To provide filtering mechanism and functionality to retrieve filtered parameters

=item To depend on other modules as least as possible (complex validators and filters depending on other modules heavyly should be supplied as dependent plugin distributions)

=item To make error message formatter independent of validator

=back

This module is NOT all-in-one validation product.  This module DOES NOT provide easy configuration.  But you have to implement validation procedure with Perl code, so on such a complex condition described above, you can write codes straightforwardly, easy to understand.

=head1 LICENSE

Copyright (C) ITO Nobuaki.

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=head1 AUTHOR

ITO Nobuaki E<lt>dayflower@cpan.orgE<gt>

=cut

